trace/rate_limit.go

```diff
@@ -0,0 +1,294 @@
+func (r *SharedRateLimit) AllowOther() bool {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	return r.SampleIntervalActive
+}
+
+type requestKey struct {
+	StreamID       string
+	SequenceNumber int
+}
+
+type rateLimitCollector struct {
+	// Shared rate limit across all collectors
+	// (typically one is created per interface, and for outgoing vs. incoming)
+	RateLimit *SharedRateLimit
+
+	// Next collector in stack
+	NextCollector Collector
+
+	// Map of unmatched request arrival times
+	RequestArrivalTimes map[requestKey]time.Time
+
+	// Channel from RateLimit for epoch starts
+	epochCh chan time.Time
+}
+
+func (r *SharedRateLimit) NewCollector(next Collector) Collector {
+	c := &rateLimitCollector{
+		RateLimit:           r,
+		NextCollector:       next,
+		RequestArrivalTimes: make(map[requestKey]time.Time),
+		epochCh:             make(chan time.Time, 1),
+	}
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.children = append(r.children, c)
+	return c
+}
+
+func (r *SharedRateLimit) onCollectorClose(closed *rateLimitCollector) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	for i, c := range r.children {
+		if c == closed {
+			r.children = append(r.children[:i], r.children[i+1:]...)
+			return
+		}
+	}
+}
+
+func (r *rateLimitCollector) Process(pnt akinet.ParsedNetworkTraffic) error {
+	switch c := pnt.Content.(type) {
+	case akinet.HTTPRequest:
+		if r.RateLimit.AllowHTTPRequest() {
+			// Collect request and the matching response as well.
+			r.NextCollector.Process(pnt)
+			key := requestKey{c.StreamID.String(), c.Seq}
+			r.RequestArrivalTimes[key] = pnt.ObservationTime
+		}
+	case akinet.HTTPResponse:
+		// Collect iff the request is in our map. (This means responses to calls
+		// before the sampling interval will not be captured.)
+		key := requestKey{c.StreamID.String(), c.Seq}
+		if _, ok := r.RequestArrivalTimes[key]; ok {
+			delete(r.RequestArrivalTimes, key)
+			r.NextCollector.Process(pnt)
+		}
+	default:
+		if r.RateLimit.AllowOther() {
+			r.NextCollector.Process(pnt)
+		}
+	}
+
+	// Check for new epoch, in a nonblocking way
+	select {
+	case epochStart := <-r.epochCh:
+		r.expireRequests(epochStart)
+	default:
+		break
+	}
+
+	return nil
+}
+
+func (r *rateLimitCollector) Close() error {
+	// Remove self from future epoch updates
+	r.RateLimit.onCollectorClose(r)
+	return r.NextCollector.Close()
+}
+
+// expire requests that came in before the threshold value
+func (r *rateLimitCollector) expireRequests(threshold time.Time) {
+	expired := 0
+	for k, v := range r.RequestArrivalTimes {
+		if v.Before(threshold) {
+			delete(r.RequestArrivalTimes, k)
+			expired += 1
+		}
+	}
+	printer.Debugf("Expired %v old requests\n", expired)
+}
+
+func NewRateLimit(witnessesPerMinute float64) *SharedRateLimit {
+	witnessLimit := witnessesPerMinute * viper.GetDuration(RateLimitEpochTime).Minutes()
+	if witnessLimit < 1 {
+		printer.Warningln("Witnesses per minute rate is too low, rounding up to 1 per 5 minutes.")
+		witnessLimit = 1
+	}
```

Comment:
Doesn't this round up to one witness per `viper.GetDuration(RateLimitEpochTime)`?

Reply:
Yes, that's correct-- I didn't think it was worth trying to nicely format the configured value (or introduce the term "epoch".)